src/S3/S3Transfer/S3TransferManager.php

```diff
@@ -802,4 +806,210 @@ private function resolvesOutsideTargetDirectory(
+        if (is_null($progressTracker)
+            && ($config['track_progress'] ?? $this->config->isTrackProgress())
+        ) {
+            $progressTracker = new SingleProgressTracker();
+        }
+        if ($progressTracker !== null) {
+            $listeners[] = $progressTracker;
+        }
+        $notifier = new TransferListenerNotifier($listeners);
+        $threshold = $config['multipart_copy_threshold_bytes']
+            ?? $this->config->getMultipartUploadThresholdBytes();
+        if ($threshold < AbstractMultipartUploader::PART_MIN_SIZE) {
+            throw new InvalidArgumentException(
+                "The provided config `multipart_copy_threshold_bytes`"
+                ." must be greater than or equal to "
+                . AbstractMultipartUploader::PART_MIN_SIZE
+            );
+        }
+
+        if ($this->requiresMultipartCopy($request->getSource(), $threshold)) {
+            $mpConfig = [
+                'part_size' => $config['part_size'] ?? $this->config->getTargetPartSizeBytes(),
+                'concurrency' => $config['concurrency'] ?? $this->config->getConcurrency(),
+            ];
+
+            return $this->tryMultipartCopy(
+                $request->getSource(),
+                $request->getCopyRequestArgs(),
+                $mpConfig,
+                $notifier
+            );
+        }
+        return $this->trySingleCopy(
+            $request->getSource(),
+            $request->getCopyRequestArgs(),
+            $notifier
+        );
+    }
+
+    /**
+     * @param array $source
+     * @param array $copyRequestArgs
+     * @param array $config
+     * @param TransferListenerNotifier|null $listenerNotifier
+     * @return PromiseInterface
+     */
+    public function tryMultipartCopy(
+        array $source,
+        array $copyRequestArgs,
+        array $config = [],
+        ?TransferListenerNotifier $listenerNotifier = null
+    ): PromiseInterface
+    {
+        $createMultipartArgs = [...$copyRequestArgs];
+
+        $copier = new MultipartCopier(
+            s3Client: $this->s3Client,
+            requestArgs: $createMultipartArgs,
+            config: $config,
+            source: $source,
+            listenerNotifier: $listenerNotifier
+        );
+
+        return $copier->copy();
+    }
+
+    /**
+     * @param array $source
+     * @param array $copyRequestArgs
+     * @param ?TransferListenerNotifier $listenerNotifier
+     * @return PromiseInterface
+     * @throws S3TransferException
+     */
+    private function trySingleCopy(
+        array $source,
+        array $copyRequestArgs,
+        ?TransferListenerNotifier $listenerNotifier = null
+    ): PromiseInterface
+    {
+        $params = [
+            'Bucket' => $copyRequestArgs['Bucket'],
+            'CopySource' => $this->getSourcePath(source: $source),
+            'Key' => $copyRequestArgs['Key']
+        ];
+
+        $objectSize = $this->s3Client->headObject([
```

Comment:
This should be wrapped in a try/catch and probably extracted to its own dedicated method like `getSourceSize()` on `MultipartCopy`